File: aicq-server/src/api/groupRoutes.ts
import { Router, Request, Response } from 'express';
import { generalLimiter } from '../middleware/rateLimit';
import * as groupService from '../services/groupService';
import { store } from '../db/memoryStore';

const router = Router();

// ─── 创建群组 ─────────────────────────────────────────────────

/**
 * POST /api/v1/group/create
 * 创建新群组
 */
router.post('/group/create', generalLimiter, (req: Request, res: Response) => {
  try {
    const { name, ownerId, description } = req.body;

    if (!name || !ownerId) {
      res.status(400).json({ error: '缺少必填字段: name, ownerId' });
      return;
    }

    const group = groupService.createGroup(name, ownerId, description);
    res.status(201).json(groupService.serializeGroup(group));
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 获取群组列表 ──────────────────────────────────────────────

/**
 * GET /api/v1/group/list?accountId=xxx
 * 获取账号所在的所有群组
 */
router.get('/group/list', generalLimiter, (req: Request, res: Response) => {
  try {
    const accountId = req.query.accountId as string;
    if (!accountId) {
      res.status(400).json({ error: '缺少查询参数: accountId' });
      return;
    }

    const groups = groupService.getGroupsForAccount(accountId);
    res.json({
      groups: groups.map(groupService.serializeGroup),
      count: groups.length,
    });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// ─── 获取群组详情 ──────────────────────────────────────────────

/**
 * GET /api/v1/group/:groupId?accountId=xxx
 * 获取群组信息（需是成员）
 */
router.get('/group/:groupId', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const accountId = req.query.accountId as string;

    if (!accountId) {
      res.status(400).json({ error: '缺少查询参数: accountId' });
      return;
    }

    const group = groupService.getGroup(groupId);
    if (!group) {
      res.status(404).json({ error: '群组不存在' });
      return;
    }

    // 检查是否是成员
    if (!groupService.isMember(groupId, accountId)) {
      res.status(403).json({ error: '你不是该群成员' });
      return;
    }

    res.json(groupService.serializeGroup(group));
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// ─── 邀请成员 ──────────────────────────────────────────────────

/**
 * POST /api/v1/group/:groupId/invite
 * 邀请成员加入群组
 */
router.post('/group/:groupId/invite', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId, targetId, displayName } = req.body;

    if (!accountId || !targetId) {
      res.status(400).json({ error: '缺少必填字段: accountId, targetId' });
      return;
    }

    const member = groupService.inviteMember(groupId, accountId, targetId, displayName);
    res.json({ success: true, member });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 踢出成员 ──────────────────────────────────────────────────

/**
 * POST /api/v1/group/:groupId/kick
 * 踢出群成员
 */
router.post('/group/:groupId/kick', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId, targetId } = req.body;

    if (!accountId || !targetId) {
      res.status(400).json({ error: '缺少必填字段: accountId, targetId' });
      return;
    }

    groupService.kickMember(groupId, accountId, targetId);
    res.json({ success: true });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 退出群组 ──────────────────────────────────────────────────

/**
 * POST /api/v1/group/:groupId/leave
 * 成员退出群组
 */
router.post('/group/:groupId/leave', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId } = req.body;

    if (!accountId) {
      res.status(400).json({ error: '缺少必填字段: accountId' });
      return;
    }

    const result = groupService.leaveGroup(groupId, accountId);
    res.json({ success: true, ...result });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 解散群组 ──────────────────────────────────────────────────

/**
 * DELETE /api/v1/group/:groupId
 * 解散群组（仅群主）
 */
router.delete('/group/:groupId', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId } = req.body;

    if (!accountId) {
      res.status(400).json({ error: '缺少必填字段: accountId' });
      return;
    }

    groupService.disbandGroup(groupId, accountId);
    res.json({ success: true });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 更新群组信息 ──────────────────────────────────────────────

/**
 * PUT /api/v1/group/:groupId
 * 更新群组名称/描述/头像
 */
router.put('/group/:groupId', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId, name, description, avatar } = req.body;

    if (!accountId) {
      res.status(400).json({ error: '缺少必填字段: accountId' });
      return;
    }

    const updated = groupService.updateGroup(groupId, accountId, { name, description, avatar });
    res.json({ success: true, group: groupService.serializeGroup(updated) });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 转让群主 ──────────────────────────────────────────────────

/**
 * POST /api/v1/group/:groupId/transfer
 * 转让群主身份
 */
router.post('/group/:groupId/transfer', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId, targetId } = req.body;

    if (!accountId || !targetId) {
      res.status(400).json({ error: '缺少必填字段: accountId, targetId' });
      return;
    }

    const group = groupService.transferOwnership(groupId, accountId, targetId);
    res.json({ success: true, group: groupService.serializeGroup(group) });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 设置成员角色 ──────────────────────────────────────────────

/**
 * POST /api/v1/group/:groupId/role
 * 设置群成员角色（仅群主）
 */
router.post('/group/:groupId/role', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId, targetId, role } = req.body;

    if (!accountId || !targetId || !role) {
      res.status(400).json({ error: '缺少必填字段: accountId, targetId, role' });
      return;
    }

    if (!['admin', 'member'].includes(role)) {
      res.status(400).json({ error: 'role 只能是 admin 或 member' });
      return;
    }

    const member = groupService.setMemberRole(groupId, accountId, targetId, role);
    res.json({ success: true, member });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 禁言/解禁成员 ─────────────────────────────────────────────

/**
 * POST /api/v1/group/:groupId/mute
 * 禁言或解禁群成员
 */
router.post('/group/:groupId/mute', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const { accountId, targetId, muted } = req.body;

    if (!accountId || !targetId || muted === undefined || muted === null) {
      res.status(400).json({ error: '缺少必填字段: accountId, targetId, muted' });
      return;
    }

    const member = groupService.muteMember(groupId, accountId, targetId, Boolean(muted));
    res.json({ success: true, member });
  } catch (err: any) {
    res.status(400).json({ error: err.message });
  }
});

// ─── 群消息历史 ──────────────────────────────────────────────

/**
 * GET /api/v1/group/:groupId/messages
 * 获取群组消息历史（支持分页）
 */
router.get('/group/:groupId/messages', generalLimiter, (req: Request, res: Response) => {
  try {
    const { groupId } = req.params;
    const accountId = req.query.accountId as string;
    const limit = Math.min(parseInt(req.query.limit as string) || 50, 200);
    const before = parseInt(req.query.before as string) || Date.now();

    if (!accountId) {
      res.status(400).json({ error: '缺少查询参数: accountId' });
      return;
    }

    const group = groupService.getGroup(groupId);
    if (!group) {
      res.status(404).json({ error: '群组不存在' });
      return;
    }

    // 检查是否是成员
    if (!groupService.isMember(groupId, accountId)) {
      res.status(403).json({ error: '你不是该群成员' });
      return;
    }

    const messages = store.groupMessages.get(groupId) || [];

    // 过滤：获取 before 之前的消息，倒序排列
    const filtered = messages
      .filter((m) => m.timestamp < before)
      .sort((a, b) => b.timestamp - a.timestamp)
      .slice(0, limit);

    res.json({
      messages: filtered,
      count: filtered.length,
      hasMore: messages.filter((m) => m.timestamp < filtered[filtered.length - 1]?.timestamp).length > 0,
    });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

export default router;

File: server/routes/group_routes.py
"""
AICQ Group Routes
===================
Group route handlers: creation, listing, detail, membership management
(invite, kick, leave, disband), ownership transfer, role management,
muting, and message history.

All routes are prefixed with ``/api/v1/group``.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, List, Optional

from aiohttp import web

from routes.middleware import (
    error_response,
    get_db,
    json_response,
)

logger = logging.getLogger("aicq.group_routes")


# ─── Group CRUD ─────────────────────────────────────────────────────────


async def create_group(request: web.Request) -> web.Response:
    """POST /api/v1/group/create  (JWT auth required)

    Request body::

        {
            "name": "My Group",
            "description": "A test group",   // optional
            "maxMembers": 100                 // optional
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    name = body.get("name")
    description = body.get("description")
    max_members = body.get("maxMembers", 100)

    if not name:
        return error_response(
            "Missing required field: name",
            status=400,
            code="MISSING_FIELDS",
        )

    owner_id = account.get("sub")

    try:
        from services.group_service import create_group as _create_group

        result = await _create_group(db, owner_id, name, description, max_members)
        return json_response(result, status=201)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error creating group")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def list_groups(request: web.Request) -> web.Response:
    """GET /api/v1/group/list  (JWT auth required)

    Returns all groups the authenticated user is a member of.
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    account_id = account.get("sub")

    try:
        from services.group_service import list_groups as _list_groups

        groups = await _list_groups(db, account_id)
        return json_response({"groups": groups}, status=200)
    except Exception as exc:
        logger.exception("Error listing groups")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def get_group(request: web.Request) -> web.Response:
    """GET /api/v1/group/{group_id}  (JWT auth required)

    Returns group detail including members list.
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]

    try:
        from services.group_service import get_group as _get_group

        group = await _get_group(db, group_id)
        if group is None:
            return error_response("Group not found", status=404, code="NOT_FOUND")
        return json_response(group, status=200)
    except Exception as exc:
        logger.exception("Error getting group")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def update_group(request: web.Request) -> web.Response:
    """PUT /api/v1/group/{group_id}  (JWT auth required)

    Request body::

        {
            "name": "Updated Name",          // optional
            "description": "New desc",       // optional
            "maxMembers": 200,               // optional
            "avatar": "url_to_avatar"        // optional
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    owner_id = account.get("sub")

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    try:
        from services.group_service import update_group as _update_group

        result = await _update_group(db, group_id, owner_id, body)
        if result is None:
            return error_response("Group not found", status=404, code="NOT_FOUND")
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error updating group")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def disband_group(request: web.Request) -> web.Response:
    """DELETE /api/v1/group/{group_id}  (JWT auth required)

    Disbands a group. Only the owner can disband.
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    owner_id = account.get("sub")

    try:
        from services.group_service import disband_group as _disband_group

        deleted = await _disband_group(db, group_id, owner_id)
        if not deleted:
            return error_response("Group not found", status=404, code="NOT_FOUND")
        return json_response({"disbanded": True, "groupId": group_id}, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error disbanding group")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


# ─── Membership Management ──────────────────────────────────────────────


async def invite_member(request: web.Request) -> web.Response:
    """POST /api/v1/group/{group_id}/invite  (JWT auth required)

    Request body::

        {
            "accountId": "abc123"
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    inviter_id = account.get("sub")

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    account_id = body.get("accountId")

    if not account_id:
        return error_response(
            "Missing required field: accountId",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.group_service import invite_member as _invite_member

        result = await _invite_member(db, group_id, account_id, inviter_id)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error inviting member")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def kick_member(request: web.Request) -> web.Response:
    """POST /api/v1/group/{group_id}/kick  (JWT auth required)

    Request body::

        {
            "accountId": "abc123"
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    kicker_id = account.get("sub")

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    account_id = body.get("accountId")

    if not account_id:
        return error_response(
            "Missing required field: accountId",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.group_service import kick_member as _kick_member

        kicked = await _kick_member(db, group_id, account_id, kicker_id)
        return json_response({"kicked": True, "accountId": account_id}, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error kicking member")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def leave_group(request: web.Request) -> web.Response:
    """POST /api/v1/group/{group_id}/leave  (JWT auth required)

    The authenticated user leaves the group. Owners cannot leave;
    they must transfer ownership or disband.
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    account_id = account.get("sub")

    try:
        from services.group_service import leave_group as _leave_group

        left = await _leave_group(db, group_id, account_id)
        return json_response({"left": True, "groupId": group_id}, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error leaving group")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def transfer_ownership(request: web.Request) -> web.Response:
    """POST /api/v1/group/{group_id}/transfer  (JWT auth required)

    Request body::

        {
            "newOwnerId": "xyz789"
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    old_owner_id = account.get("sub")

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    new_owner_id = body.get("newOwnerId")

    if not new_owner_id:
        return error_response(
            "Missing required field: newOwnerId",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.group_service import transfer_ownership as _transfer_ownership

        result = await _transfer_ownership(db, group_id, old_owner_id, new_owner_id)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error transferring ownership")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def set_member_role(request: web.Request) -> web.Response:
    """POST /api/v1/group/{group_id}/role  (JWT auth required)

    Request body::

        {
            "accountId": "abc123",
            "role": "admin"       // "admin" | "member"
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    setter_id = account.get("sub")

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    account_id = body.get("accountId")
    role = body.get("role")

    if not account_id or not role:
        return error_response(
            "Missing required fields: accountId, role",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.group_service import set_member_role as _set_member_role

        result = await _set_member_role(db, group_id, account_id, role, setter_id)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error setting member role")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def mute_member(request: web.Request) -> web.Response:
    """POST /api/v1/group/{group_id}/mute  (JWT auth required)

    Request body::

        {
            "accountId": "abc123",
            "muted": true
        }
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]
    setter_id = account.get("sub")

    try:
        body = await request.json()
    except Exception:
        return error_response("Invalid JSON body", status=400, code="INVALID_BODY")

    account_id = body.get("accountId")
    muted = body.get("muted", True)

    if not account_id:
        return error_response(
            "Missing required field: accountId",
            status=400,
            code="MISSING_FIELDS",
        )

    try:
        from services.group_service import mute_member as _mute_member

        result = await _mute_member(db, group_id, account_id, bool(muted), setter_id)
        return json_response(result, status=200)
    except ValueError as exc:
        return error_response(str(exc), status=400, code="VALIDATION_ERROR")
    except Exception as exc:
        logger.exception("Error muting member")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


async def get_group_messages(request: web.Request) -> web.Response:
    """GET /api/v1/group/{group_id}/messages  (JWT auth required)

    Query parameters::

        ?limit=50        // optional, default 50
        &before=msg_id   // optional, for cursor-based pagination

    Returns paginated message history for a group.
    """
    account = request.get("account")
    if not account:
        return error_response("Authentication required", status=401, code="AUTH_REQUIRED")

    db = get_db(request)
    group_id = request.match_info["group_id"]

    # Parse query parameters
    try:
        limit = int(request.query.get("limit", "50"))
    except (ValueError, TypeError):
        limit = 50

    before = request.query.get("before")

    try:
        from services.group_service import get_group_messages as _get_group_messages

        messages = await _get_group_messages(db, group_id, limit, before)
        return json_response({"messages": messages}, status=200)
    except Exception as exc:
        logger.exception("Error getting group messages")
        return error_response("Internal server error", status=500, code="INTERNAL_ERROR")


# ─── Route Registration ─────────────────────────────────────────────────


def setup_routes(app: web.Application, prefix: str = "/api/v1/group") -> None:
    """Register all group routes on the given aiohttp application.

    Parameters
    ----------
    app:
        The aiohttp application instance.
    prefix:
        URL prefix for all group routes. Defaults to ``/api/v1/group``.
    """
    app.router.add_post(prefix + "/create", create_group)
    app.router.add_get(prefix + "/list", list_groups)
    app.router.add_get(prefix + "/{group_id}", get_group)
    app.router.add_put(prefix + "/{group_id}", update_group)
    app.router.add_delete(prefix + "/{group_id}", disband_group)
    app.router.add_post(prefix + "/{group_id}/invite", invite_member)
    app.router.add_post(prefix + "/{group_id}/kick", kick_member)
    app.router.add_post(prefix + "/{group_id}/leave", leave_group)
    app.router.add_post(prefix + "/{group_id}/transfer", transfer_ownership)
    app.router.add_post(prefix + "/{group_id}/role", set_member_role)
    app.router.add_post(prefix + "/{group_id}/mute", mute_member)
    app.router.add_get(prefix + "/{group_id}/messages", get_group_messages)

    logger.info("Group routes registered under %s", prefix)
